src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6242,15 +6242,26 @@ private ScanTypeFlags ScanType(ParseTypeMode mode, out SyntaxToken lastTokenOfTy
                 }
             }
 
-            if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+            // Handle :: as well for error case of an alias used without a preceding identifier.
+            if (this.CurrentToken.Kind is SyntaxKind.IdentifierToken or SyntaxKind.ColonColonToken)
             {
-                result = this.ScanNamedTypePart(out lastTokenOfType);
-                if (result == ScanTypeFlags.NotType)
+                bool isAlias;
+                if (this.CurrentToken.Kind is SyntaxKind.ColonColonToken)
                 {
-                    return ScanTypeFlags.NotType;
+                    lastTokenOfType = null;
+                    result = ScanTypeFlags.NonGenericTypeOrExpression;
+                    isAlias = true;
                 }
+                else
+                {
+                    result = this.ScanNamedTypePart(out lastTokenOfType);
+                    if (result == ScanTypeFlags.NotType)
+                    {
+                        return ScanTypeFlags.NotType;
+                    }
 
-                bool isAlias = this.CurrentToken.Kind == SyntaxKind.ColonColonToken;
+                    isAlias = this.CurrentToken.Kind == SyntaxKind.ColonColonToken;
```

Comment:
>this.CurrentToken.Kind == SyntaxKind.ColonColonToken; [](start = 30, length = 53)

I understand that there is no change in behavior for this line, but I am curious. Is it possible that we skip more than just an identifier token here and stop at ```::``` token? It feels like it would be wrong to set ```isAlias``` to true in that case.

Reply:
its' a good question.  Effectively, this is a decision we make right at the start of 'scan type' (ignoring the bit about 'ref'. So this makes sense to me in teh context of:  

1. our caller is asking: do we start a type?
2. we say: if you're asking that right now, and we have a `::`, then yes, we start a type.  the same that we would answer if you asked that on `a::`

Reply:
> if you're asking that right now, and we have a ::, then yes, we start a type.  

For example, does ```A<B>::``` starts a type and uses an alias? Does it really make sense to continue with ```isAlias``` set to true in this case?

---
In reply to: [571304978](https://github.com/dotnet/roslyn/pull/51028#discussion_r571304978) [](ancestors = 571304978)

Reply:
this would only happen if ScanType was called on the `::` bit.  so it would equivalent to the behavior we would get if we had `A<B>C::`  in both cases, it should not treat it as an alias because parsing a type only allows hte coloncolon at teh start.  i'll add some tests around here to help out.

Reply:
> this would only happen if ScanType was called on the :: bit.  so it would equivalent to the behavior we would get if we had A<B>C::  in both cases, it should not treat it as an alias because parsing a type only allows hte coloncolon at teh start.  i'll add some tests around here to help out.

I am still somewhat confused by this code. My understanding is that we assume that ```result ==ScanTypeFlags.NonGenericTypeOrExpression```, let's assert this fact then. Or, perhaps, my understanding is incorrect? Since this is an existing code (unchanged in this PR), I don't think this issue is blocking, but I think this needs some follow up.  

---
In reply to: [571345168](https://github.com/dotnet/roslyn/pull/51028#discussion_r571345168) [](ancestors = 571345168)

Reply:
> let's assert this fact the

SGTM.  let me see if we can beef things up here.